src/main/java/org/mockito/internal/util/MockUtil.java

```diff
@@ -4,10 +4,14 @@
  */
 package org.mockito.internal.util;
 
+import org.mockito.MockMakers;
 import org.mockito.MockedConstruction;
 import org.mockito.Mockito;
 import org.mockito.exceptions.misusing.NotAMockException;
 import org.mockito.internal.configuration.plugins.Plugins;
+import org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker;
+import org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker;
+import org.mockito.internal.creation.proxy.ProxyMockMaker;
```

Comment:
Since we do special loading in these mockmakers, we can't load their classes until we are certain that the user wants them. Therefore, we will need to dynamically load these classes using reflection (based on the names). In other words, similar to what `DefaultMockitoPlugins` does, where we also refer to them as fully qualified names rather than the type class.

The `getMockMaker` implementation down below would then becomes something along the lines of:

```java
        Class<? extends MockMaker> type;
        switch (mockMaker) {
            case MockMakers.INLINE:
            case MockMakers.PROXY:
            case MockMakers.SUBCLASS:
                type = defaultMockitoPlugins.getDefaultPluginClass(mockMaker);
                break;
            default:
                {
                    ClassLoader loader = Thread.currentThread().getContextClassLoader();
                    if (loader == null) {
                        loader = ClassLoader.getSystemClassLoader();
                    }
                    try {
                        type = loader.loadClass(mockMaker).asSubclass(MockMaker.class);
                    } catch (Exception e) {
                        throw new IllegalStateException(
                                "Failed to load MockMaker: " + mockMaker, e);
                    }
                    break;
                }
        }
```

(Store the `defaultMockitoPlugins` in a field as a constant and make `getDefaultPluginClass` public, it's an internal class anyways so not part of our API)

If you feel adventurous, it would be good to add an integration test to ensure this remains so. I think it would be good to copy https://github.com/mockito/mockito/blob/main/src/test/java/org/mockitointegration/NoByteCodeDependenciesTest.java and adjust it to ensure that we do not load the mockmakers.

E.g. load both `Mockito.class` and `Matchers.class` and assert that it does **not** load any of the mockmakers that we ship (`.without("org.mockito.internal.creation.bytebuddy", "org.mockito.internal.creation.proxy")` on the ` ClassLoaders.excludingClassLoader()`)

Reply:
I adjusted the implementation, but I have not yet implemented a test.

I think the test you have suggested might not be sufficient. At least the JVMs which I am using do not load classes used by methods before the method is called. This means the test you have suggested succeeds with my previous implementation, at least on the JVM I am using. I think I would actually have to create a mock with another mock maker, while the built-in mock makers are not accessible. Maybe I will look at this later this week.

Reply:
Hm good point. In that case, let's punt it for now.